src/config.rs

```diff
@@ -0,0 +1,178 @@
+/// A USB vendor ID and product ID pair.
+pub struct UsbVidPid(pub u16, pub u16);
+
+macro_rules! builder_fields {
+    ( $( $(#[$meta:meta])* $name:ident: $type:ty, )* ) => {
+        $(
+            $(#[$meta])*
+            pub fn $name(mut self, $name: $type) -> Self {
+                self.$name = $name;
+                self
+            }
+        )*
+    }
+}
+
+/// A USB device configuration.
+#[derive(Clone, Copy, Debug)]
+pub struct Config<'a> {
+    /// USB device class
+    pub device_class: u8,
+    /// USB device subclass
+    pub device_sub_class: u8,
+    /// USB device protocol
+    pub device_protocol: u8,
+    /// USB control endpoint maximum packet size
+    pub max_packet_size_0: u8,
+    /// USB vendor ID
+    pub vendor_id: u16,
+    /// USB product ID
+    pub product_id: u16,
+    /// BCD encoded device release
+    pub device_release: u16,
+    /// Manufacturer string
+    pub manufacturer: Option<&'a str>,
+    /// Product string
+    pub product: Option<&'a str>,
+    /// Serial number
+    pub serial_number: Option<&'a str>,
+    /// Is device self-powered?
+    pub self_powered: bool,
+    /// Does device support remote wakeup?
+    pub supports_remote_wakeup: bool,
+    /// Is device composite with IADs?
+    pub composite_with_iads: bool,
+    /// Maximum power consumption of device
+    pub max_power: u8,
+}
```

Comment:
As much as I like the builder semantics, making struct fields public is problematic . See https://rust-lang.github.io/api-guidelines/future-proofing.html#structs-have-private-fields-c-struct-private for a discussion on why this is problematic for a maintainability perspective.

Other than that, I like the builder config API